### PR TITLE
Add custom headers to tarball fetcher

### PR DESCRIPTION
**Pull request by Glavin001**
**Summary**

My team would like to download tarballs attached to GitLab Releases.
The projects are authenticated, therefore we need to add `PRIVATE-TOKEN` header to the request for the tarball. See https://docs.gitlab.com/ee/api/README.html#personal-access-tokens

I propose the following configuration addition:

`~/.npmrc`:

```
//custom.gitlab.host/:_header:PRIVATE-TOKEN=REDACTED
```

where:
- `//custom.gitlab.host/` is the host which matches the tarball request.
- `:_header:<HEADER-NAME>` indicates a custom header should be added, with `PRIVATE-TOKEN` being the `HEADER-NAME` in this instance. Multiple custom headers could be added in this way.

Then we can run:

```
❯ yarn-local add --dev --verbose https://custom.gitlab.host/OWNER/REPO/uploads/REDACTED/PKG-NAME-v1.0.0.tgz
```

And the request will be authenticated.

Once I have approval on the above format then I will update the `CHANGELOG.md` 😃 .

**Test plan**

Once I have approval on the above format then I will add tests 😃 .

----

**Update 1**:

Now `~/.yarnrc` looks like:

```
"//custom.gitlab.host/:_header:PRIVATE-TOKEN" "REDACTED"
```

after running

```
yarn config set "//custom.gitlab.host/:_header:PRIVATE-TOKEN" "REDACTED"
```
and `~/.npmrc` is not used. Working great! 🎉

---

**Update 2**:
See https://github.com/yarnpkg/yarn/pull/6756#issuecomment-444728156

I created a release and uploaded a tarball generated from `yarn pack` to https://github.com/Glavin001/yarn/releases/tag/1.13.0-config-tarball-fetcher-headers-1

You can install this custom Yarn with:

```bash
npm install -g https://github.com/Glavin001/yarn/releases/download/1.13.0-config-tarball-fetcher-headers-1/yarn-v1.13.0-0.tgz
```

### CODE REVIEW

File: src/fetchers/tarball-fetcher.js [hunk 1]
```diff
@@ -329,3 +350,16 @@ export class LocalTarballFetcher extends TarballFetcher {
     return this.fetchFromLocal(this.reference);
   }
 }
+
+type UrlParts = {
```
Comment (Glavin001) [inline src/fetchers/tarball-fetcher.js]: This is copied from https://github.com/yarnpkg/yarn/blob/b8a565e562810f9bf86f3ac0531ad348ed7e9d0a/src/registries/npm-registry.js#L77-L88

I would like to put it in a more central location. I ask the Yarn team where would the most appropriate destination? 😄

File: src/fetchers/tarball-fetcher.js [hunk 2]
```diff
@@ -273,6 +276,24 @@ export default class TarballFetcher extends BaseFetcher {
     }
   }
 
+  requestHeaders(requestUrl: string): {[string]: string} {
```
Comment (Glavin001) [inline src/fetchers/tarball-fetcher.js]: I based this method off of https://github.com/yarnpkg/yarn/blob/b8a565e562810f9bf86f3ac0531ad348ed7e9d0a/src/registries/npm-registry.js#L200-L213

Although, looking at it now I think this `requestHeaders` does not need a `requestUrl` argument and instead could use `this.reference` directly.

**Comment by arcanis** [comment]
Given that this affect the syntax of the `.npmrc`, is it also available on npm? Or would it be a Yarn-exclusive feature? In the later case I'm a bit concerned of risks of confusion 🤔

**Comment by Glavin001** [comment]
@arcanis : Good point and I agree. This feature is *not* available on npm - I reviewed https://docs.npmjs.com/misc/config#config-settings and hopefully did not miss a similar supported feature.

I'm currently retrieving the config values with https://github.com/yarnpkg/yarn/blob/ed76774f65fa674ff3cdf3de9c6ac2de407b4235/src/fetchers/tarball-fetcher.js#L279-L281 and it appeared to only load from `.npmrc`. I'm not yet sure how to load from `.yarnrc` config file.

Any advice on how I could load config from `.yarnrc` instead of `.npmrc` would be greatly appreciated. Thank you in advance!

**Comment by arcanis** [comment]
I think you should be able to access the `yarnrc` settings through `this.config.registries.yarn` (`this.registry` is typically npm - even though it goes through the Yarn domain name).

**Comment by Glavin001** [comment]
Thank you for the assistance, @arcanis !

I pushed a commit switching from npm to yarn config.

Now `~/.yarnrc` looks like:

```
"//custom.gitlab.host/:_header:PRIVATE-TOKEN" "REDACTED"
```

after running

```bash
yarn config set "//custom.gitlab.host/:_header:PRIVATE-TOKEN" "REDACTED"
```

and `~/.npmrc` is not used. Working great! 🎉 

---

Once the above is all approved, I will add tests, push the changes, and update the Pull Request description to reflect these changes.

Thanks again for reviewing!

**Comment by Glavin001** [comment]
It looks like the Jest tests need a higher timeout:

> Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.

They fail intermittently and without related code changes.

**Comment by arcanis** [comment]
Thanks! It'll be part of the next release 😃

**Pull request merged**